src/lightning_app/runners/cloud.py

```diff
@@ -293,17 +287,93 @@ def dispatch(
         project = _get_project(self.backend.client)
 
         try:
-            list_apps_resp = self.backend.client.lightningapp_v2_service_list_lightningapps_v2(
-                project_id=project.project_id, name=app_config.name
+            if cluster_id is not None:
+                # Verify that the cluster exists
+                list_clusters_resp = self.backend.client.cluster_service_list_clusters()
+                cluster_ids = [cluster.id for cluster in list_clusters_resp.clusters]
+                if cluster_id not in cluster_ids:
+                    msg = f"You requested to run on cluster {cluster_id}, but that cluster doesn't exist."
+                    raise ValueError(msg)
+
+                self._ensure_cluster_project_binding(project.project_id, cluster_id)
+
+            # Resolve the app name, instance, and cluster ID
+            existing_instance = None
+            app_name = app_config.name
+
+            # List existing instances
+            find_instances_resp = self.backend.client.lightningapp_instance_service_list_lightningapp_instances(
+                project_id=project.project_id
             )
```

Comment:
this will break if the user has many running apps as pagination isn't handled here.